Implement channel_uses_aux() (Fixes issue #29)

SiteRelEnby opened the pull request:
Define when a channel needs to use aux LEDs as part of its operation, allowing that channel to block RGB voltage when on if configured.

Fixes: https://github.com/ToyKeeper/anduril/issues/29

ToyKeeper committed:
added initial code for emisar-d3aa torch

ToyKeeper committed:
emisar-d3aa: new model number, since this is a new product line

The 0144 model number is reserved for the successor to the Meteor M44.
This is Hank's first AA light, so it's assigned as 0161:
- 01: Emisar
- 6: product line 6
- 1: model 1

ToyKeeper committed:
d3aa: made it easy to switch between vddio2 and external voltage divider

SiteRelEnby committed:
Implement channel_uses_aux() (Fixes issue #29)

Define when a channel needs to use aux LEDs as part of its operation,
allowing that channel to block RGB voltage when on if configured.

gretel commented:
werks great for me, thanks. looks like pull request are not going to be merged though.

SiteRelEnby commented:
Not going to be merged why?

gretel commented:
> Not going to be merged why?

i dont know.

diRk262 commented:
So @gretel if pull requests doesn't get merged, how can I get the latest firmware from 2023-12-03 with this fix included? I am not able to complle it on my own 🫣

SiteRelEnby commented:
@diRk262 Which build targets do you need it for? I can build it.

SiteRelEnby commented:
[anduril-2023-12-30-channel-uses-aux.zip](https://github.com/ToyKeeper/anduril/files/13800289/anduril-2023-12-30-channel-uses-aux.zip)
@diRk262 here's all current builds (from this branch current state). Most untested, `include <disclaimer.h>`, etc, I tested on kr4 and emisar-2ch but no real reason I can think of it'd break on any weird configs.

diRk262 commented:
> [anduril-2023-12-30-channel-uses-aux.zip](https://github.com/ToyKeeper/anduril/files/13800289/anduril-2023-12-30-channel-uses-aux.zip) @diRk262 here's all current builds (from this branch current state). Most untested, `include <disclaimer.h>`, etc, I tested on kr4 and emisar-2ch but no real reason I can think of it'd break on any weird configs.

I updated my TS11 with this version. The rgb as multichannel is now working without getting overwritten by the voltage rgb function. 
No bugs found in a short test.
Thanks!

gretel commented:
@SiteRelEnby the workflow can build binaries (artifacts) if you approve it to run.

SiteRelEnby commented:
@gretel Need to be a maintainer though. Although I guess could have on my fork. Ah well, it's 2 commands to run manually anyway.

gretel commented:
> @gretel Need to be a maintainer though. Although I guess could have on my fork. Ah well, it's 2 commands to run manually anyway.

oh sorry, nothing i can do about that.

zaziboo commented:
> [anduril-2023-12-30-channel-uses-aux.zip](https://github.com/ToyKeeper/anduril/files/13800289/anduril-2023-12-30-channel-uses-aux.zip) @diRk262 here's all current builds (from this branch current state). Most untested, `include <disclaimer.h>`, etc, I tested on kr4 and emisar-2ch but no real reason I can think of it'd break on any weird configs.

Thank you for this. 

But in the TS11, it broke batt check when using the aux. Main and green channels can still blink the voltage but the rest of the colors just remain solid. I also have a TS25 if you need me to test on it.

jim-p committed:
Add Fireflies PL47G2-219 variant.

Same as PL47G2 but with FET PWM levels safe for 219 emitters.

Same as difference between PL47 and PL47-219.

PL47G2 base allows low mode for aux LEDs, PL47 does not.

SiteRelEnby committed:
Merge branch 'main' into issue29-channel-uses-aux

SiteRelEnby committed:
Prevent selecting aux channels in strobes needing brightness control

This prevents a user from 3Cing into an aux channel while in one of
these strobes (candle mode and bike flasher) where they will just get
solid aux as a result, while still allowing aux channels to be used
for strobes where it might make sense (e.g. lightning, tac strobe).

Also added uses_aux to the TS10v1 aux channel which I forgot before.

SiteRelEnby commented:
> > [anduril-2023-12-30-channel-uses-aux.zip](https://github.com/ToyKeeper/anduril/files/13800289/anduril-2023-12-30-channel-uses-aux.zip) @diRk262 here's all current builds (from this branch current state). Most untested, `include <disclaimer.h>`, etc, I tested on kr4 and emisar-2ch but no real reason I can think of it'd break on any weird configs.
> 
> Thank you for this.
> 
> But in the TS11, it broke batt check when using the aux. Main and green channels can still blink the voltage but the rest of the colors just remain solid. I also have a TS25 if you need me to test on it.

Sorry, missed this comment. I'll take a look.

There's no fundamental config differences between the TS11 and TS25 that would matter.

SiteRelEnby committed:
fix two separate bugs around RGB aux channels

First off, green was not defined properly and was instead defined the
same as a main channel (leading to the behaviour described in the issue
at https://github.com/ToyKeeper/anduril/pull/33#issuecomment-1891341281

Second, fix the actual bug causing aux to not work for blinky modes.
Specifically, set_level needs to be patched so that instead of (or as
well as, doesn't really matter) calling set_level_zero(), it will also
check for if the active channel is an RGB channel and if so them clear
RGB state.

SiteRelEnby committed:
Fix another aux channel bug

set_level_zero was causing aux-based channels to be set to off at level
0 (i.e. level 1). Also added a clarifying comment.

SiteRelEnby commented:
@zaziboo should be fixed now.

zaziboo commented:
Awesome! Thank you. 

But, unfortunately, I have no way to build hex's for now. Any chance you have the updated fixes hosted somewhere?

ToyKeeper committed:
avr32dd20-devkit: use unimplemented "CH" pin as a boost bypass control

For 3V LEDs on a boost/buck dual fuel driver, this allows
routing power around the boost converter in li-ion mode,
to reduce resistance and increase maximum output.
The AA/NiMH mode is unaffected, and boosts as normal.

SiteRelEnby committed:
Bugfix: Tactical mode has a dependency on momentary mode

Use case 1: Updating a t85 light, where tactical mode is useful but
takes some squeezing, and momentary is less useful

Use case 2: Custom builds for people who don't like/want momentary mode

ToyKeeper committed:
thefreeman-avr32dd20-devkit: enable boost bypass at boot, so AA/NiMH can boot

ToyKeeper committed:
enabled eeprom read/write stabilization by default on recent MCUs,
to prevent corrupt data

This was enabled on more build targets before, but the *_ENABLE_PIN
definition names changed and the eeprom code didn't, so it was not
compiled in for a few months on some lights which probably needed it.
To fix that, it's now enabled by default on everything bigger than
attiny85, since it doesn't really hurt anything if it's enabled when
not needed.

aslotnick committed:
use low aux for chan-aux level 0

SiteRelEnby committed:
Bugfix: Prevent switching channel modes when in tactical mode

Fixes issue https://github.com/ToyKeeper/anduril/issues/40

SiteRelEnby committed:
Allow manually running GitHub actions workflows

https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow

ToyKeeper committed:
d3aa: fixed voltage measurement

ToyKeeper committed:
d3aa fine-tuning:
- new ramp
- production style config defaults (simple mode, Hank config)
- candle tuning
- fixed way-too-fast thermal regulation
  (might still be a bit fast, but it's a lot better)

ToyKeeper committed:
fixed inaccurate comment (thanks to xikteny for spotting it)

ToyKeeper committed:
dammit, got alkaline detection half working and then my flashing adapter died
(saving progress here so I can work on a different branch)

ToyKeeper committed:
d3aa: got weak battery detection actually working,
and not letting the magic smoke out of updi adapters any more (probably)

The alkaline detection might be a little too lenient though; it
could potentially fail to activate limits when the cell is completely
full or stronger than an average alkaline.  One of my test cells
measured at 72 / 75, so if it was just a little stronger it'd pass...
but most alkalines I tried were in the 40 to 60 range and failed easily.

OTOH, if I make it easier to fail, it's likely to trip on normal li-ion
cells, and I don't want that.

So as a future enhancement idea, maybe it should have a smaller sag
threshold for AA and a larger threshold for li-ion.  That would reduce
false negatives for AA, while still preventing false positives for li-ion.

ToyKeeper committed:
weak battery detection: use different thresholds for AA and Li-Ion
(also, fixed bug where a totally empty li-ion didn't get limited)

ToyKeeper committed:
increased voltage precision from 0.025V to 0.02V
(so 0 to 255 now goes from 0.00V to 5.10V)

ToyKeeper committed:
fixed Tactical Mode's strobes when Momentary Mode not enabled

ToyKeeper committed:
use smooth steps in lockout mode, if enabled

Based on [SammysHP's patch](https://github.com/ToyKeeper/anduril/pull/18),
but only for lockout mode, not tactical.

ToyKeeper committed:
d3aa weak battery test: blink 3x instead of 2x, and omit number readout

ToyKeeper committed:
d3aa: fixed voltage calculation to use 0.02V units instead of 0.025V

ToyKeeper committed:
make-release.sh should use version-string.sh instead of duplicating code

ToyKeeper committed:
MODELS: added emisar-d3aa

ToyKeeper committed:
Forgot to update model count after the last couple additions

ToyKeeper committed:
added change log for 2024-04-20 release

ToyKeeper committed:
include hardware-specific readme files in the release .zip

SiteRelEnby committed:
Updating PR branch to rebase on latest changes, and also save a bit of
MCU space on a few lights

SiteRelEnby commented:
@zaziboo sorry, missed your comment. Just rebased it on the the latest changes, here's a zip if you want to test.
[be0361a.zip](https://github.com/ToyKeeper/anduril/files/15145812/be0361a.zip)

SiteRelEnby committed:
Merge branch 'main' into issue29-channel-uses-aux
